app/board.js

```diff
@@ -1,53 +1,97 @@
 import { Enum } from './enum'
-import { Players } from './players'
+import { Players } from './enum'
 
 class Board {
 
-  constructor() {
-    this.cells = new Array(Enum.BOARD_ROWS_NUMBER)
-    this.cleanBoard()
-    this.currentPlayer = Players.NONE
-  }
+  constructor(cells, player, size) {
+    if (cells == null){
+      let _emptyMapObj = {}
+      for (let i=0; i<Enum.BOARD_COLUMNS_NUMBER; i++)
+      for (let j=0; j<Enum.BOARD_ROWS_NUMBER; j++) {
+        let key = this.formatKey(i,j)
+        _emptyMapObj[key] = Players.NONE
+      }
+      this.cells = new Immutable.Map(_emptyMapObj)
+    } else {
+      this.cells = cells
+    }
 
+    let _currentPlayer = player
+    if (player == null) {
+      _currentPlayer = Math.round(Math.random())
+    }
+    this.currentPlayer = () => { return _currentPlayer }
 
-  cleanBoard() {
-    for(let i=0; i<Enum.BOARD_ROWS_NUMBER; i++) {
-      this.cells[i] = new Array(Enum.BOARD_COLUMNS_NUMBER)
-      for(let j=0; j<Enum.BOARD_COLUMNS_NUMBER; j++) {
-        this.cells[i][j] = Players.NONE
-      }
+    let _size = size
+    if (size == null) {
+      _size = 0
     }
+    this.size = () => { return _size }
   }
 
   forEach(func) {
-    for(let i=0; i<Enum.BOARD_ROWS_NUMBER; i++) {
-      for(let j=0; j<Enum.BOARD_COLUMNS_NUMBER; j++) {
-        func(this.cells[i][j], i, j)
-      }
+    for(let i=0; i<Enum.BOARD_ROWS_NUMBER; i++)
+    for(let j=0; j<Enum.BOARD_COLUMNS_NUMBER; j++) {
+      let key = this.formatKey(i, j)
+      func(this.cells.get(key), i, j)
     }
   }
 
-  set(row, column, player) {
-    this.cells[row][column] = player;
-    // TODO checkWin()
-    this.switchPlayer()
-  }
-
-  switchPlayer() {
-    switch (this.currentPlayer) {
+  switchPlayer(player) {
+    switch (player) {
       case Players.ONE:
-        this.currentPlayer = Players.TWO
+        player = Players.TWO
         break
       case Players.TWO:
-        this.currentPlayer = Players.ONE
+        player = Players.ONE
         break
       default:
-        this.currentPlayer = Math.round(Math.random())
+        player = Math.round(Math.random())
     }
+    return player
   }
 
-  get(row, column, func) {
-    func(this.cells[row][column])
+  addPiece(x, y) {
+    let key = this.formatKey(x, y)
+    if (this.cells.get(key) == Players.NONE) {
+      let player = this.currentPlayer()
+      let newCells = this.cells.set(key, player)
+      let newPlayer = this.switchPlayer(player)
+      let newSize = this.size()+1
+      return new Board(newCells, newPlayer, newSize)
+    } else {
+      return this
+    }
```

Comment:
nao precisa desse `else` redundante

`...`
`return new Board(newCells, newPlayer, newSize`
`}`
`return this`